pandas/core/reshape/merge.py

```diff
@@ -1678,21 +1678,19 @@ def _convert_to_multiindex(index: Index) -> MultiIndex:
     join_codes = join_index.codes
     join_names = join_index.names
 
-    # lindexer and rindexer hold the indexes where the join occurred
-    # for left and right respectively. If left/right is None then
-    # the join occurred on all indices of left/right
-    if lindexer is None:
-        lindexer = range(left.size)
-
-    if rindexer is None:
-        rindexer = range(right.size)
-
     # Iterate through the levels that must be restored
     for dropped_level_name in dropped_level_names:
+        # lindexer and rindexer hold the indexes where the join occurred
+        # for left and right respectively. If left/right is None then
+        # the join occurred on all indices of left/right
         if dropped_level_name in left.names:
+            if lindexer is None:
```

Comment:
Why are you pulling this inside the loop? This is harder to read than before

Reply:
Agreed. We can actually avoid calling take_nd when the indexers are None as that means "take everything". I made another commit which further improves times and should be clearer.

```
       before           after         ratio
     [a712c501]       [ad3f42b5]
     <main>           <multiindex-join-subset>
-    43.8±1ms         10.5±1ms        0.24  join_merge.JoinMultiindexSubset.time_join_multiindex_subset
```